### ISSUE TITLE
[Bug]: StructuredPlannerAgent always defaults to single task plan.

### ISSUE THREAD

**Issue by jabbasj** [run-llama/llama_index#16625]
### Bug Description

StructuredPlannerAgent doesn't seem to be creating sub-tasks :(

I tried my own toolset and wasn't getting any success - my setup has multiple tools: txt2sql, visualization, semantic search, etc.

Then I thought maybe my use-case was too complex so I tried a dummy example with tools = [
    product_price_tool, 
    products_available_tool, 
    product_description_tool, 
    check_availability_tool
]
and I asked things like "for each product available, get me the their prices and find me the description of the cheapest product", etc.. no sub-tasks.

Finally I gave up and tried the example from the [documentation ](https://docs.llamaindex.ai/en/stable/examples/agent/structured_planner) (see code below) because I thought maybe it didn't find simple function calls 'complex enough' and it may work with query_index better.... but still no sub-tasks (unlike what's shown in the example)... the only difference between my code and the example in the doc is that I'm using VertexAI/Gemini instead of chatgpt4/openai

It simply never creates a multi sub-task plan, I even tried changing the initial_plan_prompt and plan_refine_prompt to emphasize always creating sub-tasks no matter the complexity, etc... but no success

Btw, I do get the answers in the end but it behaves as if I just used a FunctionCallingAgent not as the StructuredPlannerAgent

### Version

0.11.18

### Steps to Reproduce

```
from llama_index.core.tools import FunctionTool
        from llama_index.core.agent import FunctionCallingAgentWorker, StructuredPlannerAgent
        from llama_index.llms.vertex import Vertex
        from llama_index.core import Settings
        from llama_index.embeddings.vertex import VertexTextEmbedding
        from llama_index.core import VectorStoreIndex, SimpleDirectoryReader
        from llama_index.core.tools import QueryEngineTool

        embed_model = VertexTextEmbedding(
            client_email=...
            token_uri=...
            private_key=...
            private_key_id=...
        )

        llm = Vertex(model='gemini-1.5-pro-002',
                    max_tokens=8000,
                    temperature=0.1)

        Settings.embed_model = embed_model
        Settings.llm = llm

        # Load documents, create tools
        lyft_documents = SimpleDirectoryReader(
            input_files=["pipelines/lyft_2021.pdf"]
        ).load_data()
        uber_documents = SimpleDirectoryReader(
            input_files=["pipelines/uber_2021.pdf"]
        ).load_data()

        lyft_index = VectorStoreIndex.from_documents(lyft_documents)
        uber_index = VectorStoreIndex.from_documents(uber_documents)

        lyft_tool = QueryEngineTool.from_defaults(
            lyft_index.as_query_engine(llm = llm),
            name="lyft_2021",
            description="Useful for asking questions about Lyft's 2021 10-K filling.",
        )

        uber_tool = QueryEngineTool.from_defaults(
            uber_index.as_query_engine(llm = llm),
            name="uber_2021",
            description="Useful for asking questions about Uber's 2021 10-K filling.",
        )

        tools = [
            lyft_tool, 
            uber_tool, 
    
        ]

        agent_worker = FunctionCallingAgentWorker.from_tools(
            tools, llm=llm,
            verbose=True
        )


        agent = StructuredPlannerAgent(
            agent_worker, tools=tools, verbose=True, llm=llm,
        )


        message = "Summarize the key risk factors for Lyft and Uber in their 2021 10-K filings."

        plan_id = agent.create_plan(message)
        plan = agent.state.plan_dict[plan_id]

        yield f"\n{plan}"

        response = agent.chat(message)

        yield f"\n{response}"
```

### Relevant Logs/Tracbacks

```shell
No complex plan predicted. Defaulting to a single task plan.
=== Initial plan ===
default:
Summarize the key risk factors for Lyft and Uber in their 2021 10-K filings. ->
```

**Comment by jabbasj** [comment]
@dosu I did try modifying the initial_plan_prompt and plan_refine_prompts

For example:

DEFAULT_INITIAL_PLAN_PROMPT = """\
Think step-by-step. Given a task and a set of tools, create a comprehesive, end-to-end plan to accomplish the task.
YOU MUST ALWAYS GENERATE SUB-TASKS NO MATTER HOW SIMPLE THE REQUEST IS.
The plan should end with a sub-task that satisfies the overall task.

The tools available are:
{tools_str}

Overall Task: {task}
"""

Is StructuredPlannerAgent prompt-based/output-parsing dependent like ReactAgent or does it leverage function calling ? 
Is it possible the functioncalling api is putting the "plan" in the tool calls part of the response but that isn't parsed by the planner agent?

**Comment by logan-markewich** [comment]
Vertex/Gemini could be just really bad at following prompt instructions? 

If the LLM class doesn't implement `FunctionCallingLLM` base class, then it resorts to manual prompting/output parsing to parse structured outputs, which is very fragile. If it breaks, it defaults to a single task/plan
https://github.com/run-llama/llama_index/blob/227145fb94fcaa4da02d559fc81843fcb2af2b57/llama-index-core/llama_index/core/agent/runner/planner.py#L245

Looking at LLM base classes, 
- `Gemini` does not implement FunctionCallingLLM https://github.com/run-llama/llama_index/blob/227145fb94fcaa4da02d559fc81843fcb2af2b57/llama-index-integrations/llms/llama-index-llms-gemini/llama_index/llms/gemini/base.py#L52
- `Vertex` does: https://github.com/run-llama/llama_index/blob/227145fb94fcaa4da02d559fc81843fcb2af2b57/llama-index-integrations/llms/llama-index-llms-vertex/llama_index/llms/vertex/base.py#L50
    - but only if its a gemini model through vertex https://github.com/run-llama/llama_index/blob/227145fb94fcaa4da02d559fc81843fcb2af2b57/llama-index-integrations/llms/llama-index-llms-vertex/llama_index/llms/vertex/base.py#L178
    - which seems like it relies on the model name starting with `gemini`: https://github.com/run-llama/llama_index/blob/227145fb94fcaa4da02d559fc81843fcb2af2b57/llama-index-integrations/llms/llama-index-llms-vertex/llama_index/llms/vertex/gemini_utils.py#L9

**Comment by logan-markewich** [comment]
(If you don't have the latest vertex package too, you might want to update, `pip install -U llama-index-llms-vertex`

**Comment by jabbasj** [comment]
I am using the Vertex class and not Gemini, I only said Vertex/Gemini because I'm actually using Vertex('gemini-1.5-pro-002')
I am using llama_index-llms-vertex==0.3.7 which is the latest

I am not having any issues with function calling, I can see function calls/outputs going through just as the FunctionCallingAgent does by itself.

I also thought the model may be bad at following prompt instructions - that's why I tried to recreate the example from the documentation without having my own custom system prompts, etc... I mean I tend to agree vertex/gemini-1.5-pro isn't as good as chatgpt4 but this is a rather simple prompt with only 2 tools - so I don't think it's that bad... and just for funzies I tried doing a raw query through the vertex studio:

```
Think step-by-step. Given a task and a set of tools, create a comprehesive, end-to-end plan to accomplish the task.
Keep in mind not every task needs to be decomposed into multiple sub-tasks if it is simple enough.
The plan should end with a sub-task that satisfies the overall task.

The tools available are:
lyft_tool:"Useful for asking questions about Uber's 2021 10-K filling."
ubert_tool: "Useful for asking questions about Lyft's 2021 10-K filling."

Overall Task: Summarize the key risk factors for Lyft and Uber in their 2021 10-K filings.
```

**vertex_ai/gemini-1.5-pro**

```
Identify Lyft's risk factors: Use the lyft_tool to ask "What are the key risk factors mentioned in Lyft's 2021 10-K filing?"

Identify Uber's risk factors: Use the ubert_tool to ask "What are the key risk factors mentioned in Uber's 2021 10-K filing?"

Combine and summarize: Combine the information gathered from both tools, summarizing the key risk factors for both Lyft and Uber. This is the final step that completes the overall task.
```

So clearly the LLM does get the idea of what needs to be done... so maybe the output format is not being respected

Is there any easy way for me to log/monitor the exact payload sent to/received from the LLM?

**Comment by jabbasj** [comment]
![image](https://github.com/user-attachments/assets/7953ae32-ef76-48d6-b446-2ca397ab8c53)

So this is where it's "breaking" and defaulting to the single plan, there seems to this error `'Unknown field for Schema: $defs'`

**Comment by logan-markewich** [comment]
Seems like it needs a PR to update that function?

**Comment by jabbasj** [comment]
### The full stack trace:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/agent/runner/planner.py", line 239, in create_plan
    plan = self.llm.structured_predict(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 307, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/llms/llm.py", line 371, in structured_predict
    result = program(llm_kwargs=llm_kwargs, **prompt_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 307, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/program/function_program.py", line 202, in __call__
    agent_response = self._llm.predict_and_call(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^\\
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 307, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/llms/function_calling.py", line 182, in predict_and_call
    response = self.chat_with_tools(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/llms/function_calling.py", line 48, in chat_with_tools
    response = self.chat(**chat_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 307, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/core/llms/callbacks.py", line 173, in wrapped_llm_chat
    f_return_val = f(_self, messages, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/llms/vertex/base.py", line 240, in chat
    generation = completion_with_retry(
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/llms/vertex/utils.py", line 113, in completion_with_retry
    return _completion_with_retry(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 475, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 376, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^\\
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 398, in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^\\
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^\\
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 478, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^\\
  File "/usr/local/lib/python3.11/site-packages/llama_index/llms/vertex/utils.py", line 92, in _completion_with_retry
    tools = to_gemini_tools(tools) if tools else []
            ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/llama_index/llms/vertex/utils.py", line 60, in to_gemini_tools
    func_name = FunctionDeclaration(
                ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/vertexai/generative_models/_generative_models.py", line 1657, in __init__
    raw_schema = aiplatform_types.Schema(gapic_schema_dict)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/proto/message.py", line 724, in __init__
    raise ValueError(
ValueError: Unknown field for Schema: $defs
```

**Comment by jabbasj** [comment]
I tried upgrading `google-cloud-aiplatform` but issue persists.

Based on my understanding and the error message `'Unknown field for Schema: $defs'` it seems like $defs is not supposed to be part of 'parameters'**

```
class FunctionDeclaration:
    r"""A representation of a function declaration.

    Usage:
        Create function declaration and tool:
        '''
        get_current_weather_func = generative_models.FunctionDeclaration(
            name="get_current_weather",
            description="Get the current weather in a given location",
            parameters={
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": "The city and state, e.g. San Francisco, CA"
                    },
                    "unit": {
                        "type": "string",
                        "enum": [
                            "celsius",
                            "fahrenheit",
                        ]
                    }
                },
                "required": [
                    "location"
                ]
            },
        )
```

Below is the output of `tools[0].metadata.get_parameters_dict()` which is producing this $defs

```
def get_parameters_dict(self) -> dict:
        if self.fn_schema is None:
            parameters = {
                "type": "object",
                "properties": {
                    "input": {"title": "input query string", "type": "string"},
                },
                "required": ["input"],
            }
        else:
            parameters = self.fn_schema.model_json_schema()
            parameters = {
                k: v
                for k, v in parameters.items()
                if k in ["type", "properties", "required", "definitions", "$defs"]
            }
        return parameters
```

```
{'$defs': {'SubTask': {...}}, 'properties': {'sub_tasks': {...}}, 'required': ['sub_tasks'], 'type': 'object'}
special variables:
function variables:
'$defs': {'SubTask': {'description': 'A single sub-task in a plan.', 'properties': {...}, 'required': [...], 'title': 'SubTask', 'type': 'object'}}
special variables:
function variables:
'SubTask': {'description': 'A single sub-task in a plan.', 'properties': {'name': {...}, 'input': {...}, 'expected_output': {...}, 'dependencies': {...}}, 'required': ['name', 'input', 'expected_output', 'dependencies'], 'title': 'SubTask', 'type': 'object'}
special variables:
function variables:
'description': 'A single sub-task in a plan.'
'properties': {'name': {'description': 'The name of the sub-task.', 'title': 'Name', 'type': 'string'}, 'input': {'description': 'The input prompt for the sub-task.', 'title': 'Input', 'type': 'string'}, 'expected_output': {'description': 'The expected output of the sub-task.', 'title': 'Expected Output', 'type': 'string'}, 'dependencies': {'description': 'The sub-task names that must be completed before this sub-task.', 'items': {...}, 'title': 'Dependencies', 'type': 'array'}}
special variables:
function variables:
'name': {'description': 'The name of the sub-task.', 'title': 'Name', 'type': 'string'}
special variables:
function variables:
'description': 'The name of the sub-task.'
'title': 'Name'
'type': 'string'
len(): 3
'input': {'description': 'The input prompt for the sub-task.', 'title': 'Input', 'type': 'string'}
special variables:
function variables:
'description': 'The input prompt for the sub-task.'
'title': 'Input'
'type': 'string'
len(): 3
'expected_output': {'description': 'The expected output of the sub-task.', 'title': 'Expected Output', 'type': 'string'}
special variables:
function variables:
'description': 'The expected output of the sub-task.'
'title': 'Expected Output'
'type': 'string'
len(): 3
'dependencies': {'description': 'The sub-task names that must be completed before this sub-task.', 'items': {'type': 'string'}, 'title': 'Dependencies', 'type': 'array'}
special variables:
function variables:
'description': 'The sub-task names that must be completed before this sub-task.'
'items': {'type': 'string'}
special variables:
function variables:
'type': 'string'
len(): 1
'title': 'Dependencies'
'type': 'array'
len(): 4
len(): 4
'required': ['name', 'input', 'expected_output', 'dependencies']
'title': 'SubTask'
'type': 'object'
len(): 5
len(): 1
'properties': {'sub_tasks': {'description': 'The sub-tasks in the plan.', 'items': {...}, 'title': 'Sub Tasks', 'type': 'array'}}
special variables:
function variables:
'sub_tasks': {'description': 'The sub-tasks in the plan.', 'items': {'$ref': '#/$defs/SubTask'}, 'title': 'Sub Tasks', 'type': 'array'}
special variables:
function variables:
'description': 'The sub-tasks in the plan.'
'items': {'$ref': '#/$defs/SubTask'}
'title': 'Sub Tasks'
'type': 'array'
len(): 4
len(): 1
'required': ['sub_tasks']
'type': 'object'
len(): 4
```

I'm not sure why **$defs** is here or what its purpose is... quick google searches aren't yielding anything (I thought maybe it's a thing for OpenAI function calling API)

**Comment by jabbasj** [comment]
I tried yolo just getting rid of **$defs** and it started throwing this error instead: `('Protocol message Schema has no "$ref" field.',)` so it seems like it's not a useless field lol 

I don't think I know enough about the dependencies and inner workings to fix this but I'm happy to help if you want me to test things with Vertex @logan-markewich

**Comment by jabbasj** [comment]
@dosu any suggestions for how I could fix this bug ?

**Comment by stfines-clgx** [comment]
I'm working on this and other issues with Vertex/Gemini - the update to google's generativeai 1.70.0 version breaks a lot of things in the integration and it is taking some time for me to get to them- but this is my day job for my company so I'm going to keep plugging away at it and eventually I'll have a version of the integration that works with the current Gemini/Vertex versions. 

This particular problem is because the google library itself uses an out of date internal dependency - I've reached out to my Google support rep to see if I can get that issue resolved (or at least get a hotfix for it) but I only just hit this issue and it's late on a Friday so I don't really expect any movement today.

**Comment by stfines-clgx** [comment]
> I tried yolo just getting rid of **$defs** and it started throwing this error instead: `('Protocol message Schema has no "$ref" field.',)` so it seems like it's not a useless field lol or maybe I'm running into a separate issue?
> 
> I'm also confused why this would impact StructuredPlannerAgent and not FunctionCallingAgent, I would've thought the same/similar code is used for both
> 
> I don't think I know enough about the dependencies and inner workings to fix this but I'm happy to help if you want me to test things with Vertex @logan-markewich

the Json to Protobuf handler will treat all fields that start with $ as a protobuf extension so that's part of the issue.

**Comment by stfines-clgx** [comment]
After conversation with Google support, it is because of the way that pydantic serializes the Plan object. The internal schema object for google-cloud-aiplatform doesn't recognize the schema type $ref or $def so it throws this error. Pydantic does not provide a mechanism to rename the schema field - it provides a template value for the reference values but not the names [Pydantic Documentation](https://docs.pydantic.dev/latest/api/pydantic_core/).

I've pointed out to Google that they need to add this support to their library because it isn't possible to make this change to pydantic.

**Comment by jabbasj** [comment]
> After conversation with Google support, it is because of the way that pydantic serializes the Plan object. The internal schema object for google-cloud-aiplatform doesn't recognize the schema type $ref or $def so it throws this error. Pydantic does not provide a mechanism to rename the schema field - it provides a template value for the reference values but not the names [Pydantic Documentation](https://docs.pydantic.dev/latest/api/pydantic_core/).
> 
> I've pointed out to Google that they need to add this support to their library because it isn't possible to make this change to pydantic.

so this is not something we can monkey-patch on our end ?

do you have any idea why this issue happens for the StructuredPlannerAgent but not the FunctionCallingAgent ?

**Comment by stfines-clgx** [comment]
I haven't looked at `FunctionCallingAgent`, but I don't think it serializes a `Plan` object into it's tool call - which is what serializes with the json schema in it's `tool_parameter` field. 

I'm trying to see if I can override how the `to_gemini_tool` method creates the tool objects and perform a google-compatible serialization but finding a generic serialization that is compatible with both protobuf _and_ pydantic is challenging - and I'm not familiar enough with either to just whip one up. I'll keep plugging away - both at google and at the problem to see if I can get this figured out and into a PR.

**Comment by stfines-clgx** [comment]
This is now a [Google issue](https://issuetracker.google.com/issues/377995131) you can look follow the status at the link and star it so you get updates as it progresses. 

I'll continue to plug away at custom serialization formats to make this work - I think I found a way; but only some experimentation will tell.

**Comment by jabbasj** [comment]
> This is now a [Google issue](https://issuetracker.google.com/issues/377995131) you can look follow the status at the link and star it so you get updates as it progresses.
> 
> I'll continue to plug away at custom serialization formats to make this work - I think I found a way; but only some experimentation will tell.

Many thanks @stfines-clgx 
I [saw](https://github.com/run-llama/llama_index/compare/main...stfines-clgx:llama_index:dev/1_70_protobuf_only) you were able to adapt the multi-concierge to use vertex and fixed a related bug... did you have good success with that ?

**Comment by stfines-clgx** [comment]
I didn't have much luck with it - but I did get it to work; I also got the Planner agent in this defect to work. I'll push that to my open pr and fork tonight if you want to pull it down and test it out to give me feedback.

**Comment by jabbasj** [comment]
> I didn't have much luck with it - but I did get it to work; I also got the Planner agent in this defect to work. I'll push that to my open pr and fork tonight if you want to pull it down and test it out to give me feedback.

Hey, I got the multi-concierge example working without your fork

```
llama-index==0.11.10
llama-index-core==0.11.10
llama_index-llms-vertex==0.3.4
llama-index-utils-workflow==0.3.0
google-cloud-aiplatform==1.56.0
```

I did only minor modifications:

- Switch `llm = OpenAI(model="gpt-4o", temperature=0.4)` to `llm = Vertex(model="gemini-1.5-flash-002",max_tokens=8000,temperature=0.4)`

- Add instructions to the prompts to make sure orchestrator kicked in from the sub agents properly

- Put some slight validations in workflow.py to make it fail cleaner when `"agent_name" is not in tool_call.tool_kwargs` or if `tool = tools_by_name.get(tool_call.tool_name)` returns None

- Change the get_account_id to not always return the same user's account id

But after that, I got a nice workflow running, here's a sample output:

```
Orchestrator taking over...

SYSTEM >> Transferring to agent Authentication Agent

SYSTEM >> Recording username
SYSTEM >> Tool store_username called with {'username': 'test_username'} returned None

AGENT >> Please provide your password.


USER >> Okay, my password is  'password123'.


SYSTEM >> Logging in user test_username
SYSTEM >> Tool login called with {'password': 'password123'} returned Logged in user test_username with session token 1234567890. They have an account with id 123 and a balance of $1000.
SYSTEM >> Agent is requesting a transfer. Please h[...]
AGENT >> What can I help you with today?


USER >> Okay, I'd like to check my account balance first.  Then, I want to transfer some money.  Finally, I'd like to know the current price of Apple stock (AAPL).

Orchestrator taking over...

SYSTEM >> Transferring to agent Account Balance Agent

SYSTEM >> Checking if authenticated
SYSTEM >> Looking up account balance for 123
SYSTEM >> Tool get_account_balance called with {'account_id': '123'} returned Account 123 has a balance of $1000

SYSTEM >> Agent is requesting a transfer. Please hold.
Orchestrator taking over...

AGENT >> Okay, your balance is $1000.  Now, how much money would you like to transfer and to which account?


USER >> Okay, I want to transfer $200 to account number 456.  Then tell me the price of AAPL.

Orchestrator taking over...

AGENT >> 

USER >> Okay, the transfer is complete. Now, what's the price of AAPL?

Orchestrator taking over...

SYSTEM >> Transferring to agent Stock Lookup Agent

SYSTEM >> Looking up stock price for AAPL
SYSTEM >> Tool lookup_stock_price called with {'stock_symbol': 'AAPL'} returned Symbol AAPL is currently trading at $100.00

AGENT >> Okay, the current price of AAPL is $100.00. Is there anything else I can help you with?

USER >> exit
```

Is this too good to be true? Is it really using the vertex function calling api ?

**Comment by xmanatee** [comment]
Faced the same issue when using local llm.
Briefly looked through this discussion and all the right things were said already.

One thing I'd argue is that "it's Google Issue" and not pydantic or llama_index issue :) The fact that pydantic doesn't allow inlining nested Models and that llama_index doesn't do much about it feels wrong.

Fixed the issue for myself with a dirty hack:
```
from llama_index.core.agent.runner.planner import Plan

def fix_model_json_schema_method(clazz):
    schema = clazz.model_json_schema()
    if not "$defs" in schema:
        return schema
    defs = schema["$defs"]
    schema = inline_model_json_schema(schema, defs)
    del schema["$defs"]

    def return_fixed_schema(cls, **kwargs):
        return schema
    
    clazz.model_json_schema = classmethod(return_fixed_schema)


def inline_model_json_schema(schema, defs) -> dict:
    if isinstance(schema, list):
        return [inline_model_json_schema(i, defs) for i in schema]

    if not isinstance(schema, dict):
        return schema

    if not len(schema) == 1 or "$ref" not in schema:
        return { k: inline_model_json_schema(v, defs) for k, v in schema.items()}

    ref = schema["$ref"]
    if not ref.startswith("#/$defs/"):
        print("Something is wrong in `inline_model_json_schema`")
        return schema
    
    ref_keys = ref[len("#/$defs/"):].split("/")
    result = defs
    for ref_key in ref_keys:
        result = result[ref_key]

    return result

fix_model_json_schema_method(Plan)
```

Basically overriding pydantic's `model_json_schema` method with inlined `$defs`.
Calling `fix_model_json_schema_method(Plan)` before running the StructuredPlannerAgent works like a charm for me.
